### PR TITLE
A block should be context-sensitive if any return statement is

### LINKED ISSUE #17586: Implicit any with overloaded function taking function-returning callback
**TypeScript Version:**  nightly (Version 2.5.0-dev.20170801)

**Code**

```ts
declare function f(n: number): void;
declare function f(cb: () => (n: number) => number): void;

f(() => n => n + 1);
```

**Expected behavior:**

No error.

**Actual behavior:**

`src/a.ts(4,9): error TS7006: Parameter 'n' implicitly has an 'any' type.`

There is no error if I remove the first overload of `f`.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #17586 for the case when the function uses a block body rather than an expression body.

### CODE REVIEW

**Comment by andy-ms** [comment]
I tried adding some caching to this through `NodeLinks`, but there was on average no improvement.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8414,8 +8414,7 @@ namespace ts {
                 }
             }
 
-            // TODO(anhans): A block should be context-sensitive if it has a context-sensitive return value.
-            return node.body.kind === SyntaxKind.Block ? false : isContextSensitive(node.body);
+            return isBlock(node.body) ? forEachReturnStatement(node.body, rs => rs.expression && isContextSensitive(rs.expression)) : isContextSensitive(node.body);
```
Comment (andy-ms) [inline src/compiler/checker.ts]: Yes, added a test.

**Comment by andy-ms** [comment]
@rbuckton Good to go?

**Comment by mhegazy** [comment]
can you run perf tests on this change and let us know if you see any regressions.

**Comment by andy-ms** [comment]
I don't see any change.

#### Monaco - node (v8.2.1, x64)

| Project     |            Baseline |             Current |             Delta |     Best |    Worst |
| ----------- | ------------------- | ------------------- | ----------------- | -------- | -------- |
| Memory used | 356,236k (±  0.01%) | 356,227k (±  0.00%) |    -9k (-  0.00%) | 356,205k | 356,259k |
| Parse Time  |    2.12s (±  0.38%) |    2.15s (±  2.08%) | +0.03s (+  1.42%) |    2.10s |    2.32s |
| Bind Time   |    0.81s (±  0.45%) |    0.85s (±  7.97%) | +0.05s (+  5.83%) |    0.79s |    1.09s |
| Check Time  |    4.36s (±  0.37%) |    4.46s (±  1.72%) | +0.10s (+  2.39%) |    4.34s |    4.66s |
| Emit Time   |    7.52s (±  6.44%) |    7.43s (±  1.50%) | -0.09s (-  1.20%) |    7.25s |    7.75s |
| Total Time  |   14.81s (±  3.38%) |   14.89s (±  1.48%) | +0.09s (+  0.59%) |   14.49s |   15.48s |

#### Monaco - tsc (x86)

| Project     |            Baseline |             Current |             Delta |     Best |    Worst |
| ----------- | ------------------- | ------------------- | ----------------- | -------- | -------- |
| Parse Time  |    1.50s (±  0.50%) |    1.50s (±  0.64%) | -0.01s (-  0.47%) |    1.47s |    1.52s |
| Bind Time   |    0.57s (±  2.69%) |    0.55s (±  2.56%) | -0.02s (-  2.98%) |    0.53s |    0.59s |
| Check Time  |    4.12s (±  1.98%) |    4.04s (±  1.57%) | -0.08s (-  1.99%) |    3.93s |    4.17s |
| Emit Time   |    7.70s (±  0.65%) |    7.63s (±  0.85%) | -0.07s (-  0.90%) |    7.50s |    7.75s |
| Total Time  |   13.89s (±  0.79%) |   13.72s (±  0.79%) | -0.17s (-  1.25%) |   13.53s |   13.99s |

#### TFS - node (v8.2.1, x64)

| Project     |            Baseline |             Current |             Delta |     Best |    Worst |
| ----------- | ------------------- | ------------------- | ----------------- | -------- | -------- |
| Memory used | 309,991k (±  0.01%) | 309,990k (±  0.01%) |    -1k (-  0.00%) | 309,933k | 310,034k |
| Parse Time  |    1.57s (±  2.92%) |    1.55s (±  1.73%) | -0.02s (-  1.53%) |    1.44s |    1.57s |
| Bind Time   |    0.66s (±  3.75%) |    0.67s (±  4.12%) | +0.01s (+  1.21%) |    0.65s |    0.78s |
| Check Time  |    3.83s (±  1.20%) |    3.81s (±  0.55%) | -0.03s (-  0.65%) |    3.77s |    3.85s |
| Emit Time   |    4.15s (±  1.67%) |    4.13s (±  1.22%) | -0.02s (-  0.51%) |    4.07s |    4.33s |
| Total Time  |   10.22s (±  1.22%) |   10.16s (±  0.55%) | -0.06s (-  0.62%) |   10.06s |   10.35s |

#### TFS - tsc (x86)

| Project     |            Baseline |             Current |             Delta |     Best |    Worst |
| ----------- | ------------------- | ------------------- | ----------------- | -------- | -------- |
| Parse Time  |    1.07s (±  1.20%) |    1.07s (±  0.64%) | -0.01s (-  0.65%) |    1.05s |    1.08s |
| Bind Time   |    0.58s (±  2.02%) |    0.58s (±  2.91%) | +0.00s (+  0.34%) |    0.55s |    0.62s |
| Check Time  |    3.33s (±  1.47%) |    3.34s (±  1.38%) | +0.01s (+  0.39%) |    3.24s |    3.42s |
| Emit Time   |    4.49s (±  0.84%) |    4.50s (±  0.94%) | +0.01s (+  0.22%) |    4.44s |    4.62s |
| Total Time  |    9.47s (±  0.55%) |    9.49s (±  0.85%) | +0.02s (+  0.19%) |    9.31s |    9.66s |

**Comment by andy-ms** [comment]
@mhegazy Good to go?

**Comment by andy-ms** [comment]
@mhegazy @rbuckton Good to go?

**Comment by andy-ms** [comment]
@mhegazy @rbuckton

**Comment by andy-ms** [comment]
@mhegazy @rbuckton

**Comment by mhegazy** [comment]
I am reluctant to take this one. it is definitely more work, and so far there has been no real-world scenario that is affected; on the other hand it is a correct change..

**Comment by andy-ms** [comment]
OK, we can bring this back if we see someone complain.

**Pull request closed without merging**